### PR TITLE
SRE-35 haskell overhaul

### PR DESCRIPTION
**Pull request by craigem**
This is the overhaul of Haskell for `cardano-ledger` along the lines of `cardano-node` [PR #507](https://github.com/input-output-hk/cardano-node/pull/507/).

### CODE REVIEW

**Comment by nc6** [comment]
@craigem Is this ready for review? There are a bunch of commented lines, and CI is failing.

File: scripts/nix/fetch-nixpkgs.nix [hunk 1]
```diff
@@ -0,0 +1,6 @@
+let
+  spec = builtins.fromJSON (builtins.readFile ./nixpkgs-src.json);
+in builtins.fetchTarball {
```
Comment (cleverca22) [inline scripts/nix/fetch-nixpkgs.nix]: this this be replaced by niv?

File: nix/default.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{ system ? builtins.currentSystem
+, crossSystem ? null
+, config ? {}
+, sourcesOverride ? {}
+}:
+let
+  sources = import ./sources.nix { inherit pkgs; }
+    // sourcesOverride;
+  iohKNix = import sources.iohk-nix {};
```
Comment (nc6) [inline nix/default.nix:9]: Weird capitalisation here :-)

**Review by nc6 — APPROVED** [review]

**Review by disassembler — APPROVED** [review]

**Pull request merged**